vllm/hpu/cache_ops.py

```diff
@@ -10,35 +10,49 @@
 import habana_frameworks.torch as htorch
 
 
-def reshape_and_cache(key, value, key_cache, value_cache, slot_mapping, is_prompt=False):
-    """
-    key: [num_tokens, num_heads, head_size]
-    value: [num_tokens, num_heads, head_size]
-    key_cache: [num_heads, head_size, block_size] * num_blocks
-    value_cache: [num_heads, head_size, block_size] * num_blocks
-    slot_mapping: [num_tokens]
-    """
-    num_tokens = key.shape[0]
-    block_size = key_cache.shape[-1]
-    slot_mapping = slot_mapping.to(key.device)
-    block_indices = torch.div(slot_mapping, block_size, rounding_mode="floor")
+def pad_to_full_block(data, block_size, pad_value):
+    seq_dim = 1
+    pad_shape = list(data.shape)
+    remainder = pad_shape[seq_dim] % block_size
+    if remainder == 0:
+        return data
+    pad_shape[seq_dim] = block_size - remainder
+    pad = torch.full(pad_shape, pad_value, dtype=data.dtype, device=data.device)
+    return torch.cat([data, pad], dim=seq_dim)
+
+
+def initialize_cache(data, indices, cache):
+    block_size = cache.size(-1)
+    data = data.unflatten(0, (-1, block_size)).permute(0, 2, 3, 1)
+    indices = indices.unflatten(0, (-1, block_size))[:,0]
+    cache.index_copy_(0, indices, data)
+
+
+def update_cache(data, indices, offsets, cache):
+    prev = cache.index_select(0, indices)
+    idx = offsets.view(-1, 1, 1, 1).expand(-1, data.size(1), data.size(2), -1)
+    prev.scatter_(-1, idx, data.unsqueeze(-1))
+    cache.index_copy_(0, indices, prev)
+
+
+def reshape_and_cache(key, value, key_cache, value_cache, slot_mapping, is_prompt):
+    block_size = key_cache.size(-1)
+    assert slot_mapping.dim() == 2, 'This implementation requires unflattened slot_mapping!'
```

Comment:
Can this requirement be moved under the prompt code path to minimize API divergence?

Reply:
I understand your concern. For a moment I thought we could use number of dims to detect whether we're in prompt or in generation (flattened for generation and unflattened for prompt), but it has exactly the same problem as with using `is_prompt`. If someone forgets about it it will silently fail. Perhaps we'll come up with a better implementation later on that would allow us to use exactly the same API as nv, but for now, I'd keep the API separate.